### ISSUE TITLE
"network", "network-alias", and "name" which are longer than 64 characters make networks between containers malfunction

### ISSUE THREAD

**Issue by eubnara** [moby/moby#37971]
**Description**

Some `docker` option values longer than 64 characters make networks between containers malfunction.

1. `--network`

If I create a network with a name longer than 64 characters and create 2 containers with it, 2 containers cannot communicate each other.

2. `--name`, `--network-alias`

If I run a container with an option (`---name` or `--network-alias`) longer than 64 characters, a container which is in the same network cannot communicate with it.

**Steps to reproduce the issue:**
1. create two containers with a name longer than 64 characters.
2. make two containers in the same network.
3. Try to connect one container and ping the other container.

**Describe the results you received:**

Failed to ping the other container.

**Describe the results you expected:**

I can ping in one container to the other container with the name longer than 64 characters which is specified by `--name` or `--network-alias`

**Output of `docker version`:**

```
$ docker -v
Docker version 18.03.1-ce, build 9ee9f40
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

physical, centos 7

**Comment by olljanat** [comment]
Derek add label: version/18.03

**Comment by olljanat** [comment]
@fcrisciani this one sounds like libnetwork issue. PTAL

**Comment by thaJeztah** [comment]
I suspect this may be due to DNS having a length restriction of 63 characters per label;

https://tools.ietf.org/html/rfc1123#page-79

>    The DNS defines domain name syntax very generally -- a
>    string of labels each containing up to 63 8-bit octets,
>     separated by dots, and with a maximum total of 255
>    octets.

Not sure if the length is restricted in docker/libnetwork, or if it just fails due to the length exceeding 63 characters.

**Comment by olljanat** [comment]
@thaJeztah good point. Then when probably should prevent users to use longer network names?

**Comment by thaJeztah** [comment]
> Then when probably should prevent users to use longer network names?

That's a tricky question; for example, at some point, we implemented validation for valid environment-variable names (according to the specifications), but it turns out that Linux actually allowed invalid names, and that there's software out there that _expects_ those invalid names to work. Our validation therefore broke that software, so we decided to keep validation to a minimum, and defer handling of valid (and invalid) names to the kernel / container.

Although I generally like having things validated, the same issue could occur here. So if we currently don't actively _block_ such names, I'm not sure we can easily change that (without possibly breaking some use-case).

Having said the above; I'm not 100% sure if there's no check already somewhere (in which case possibly the issue could be related to that, if names are trimmed to 63 chars?)

**Comment by thaJeztah** [comment]
I suggest to;

1. Get to the bottom of the issue (verify if it's indeed due to the DNS limitation, or if there's something else)
2. If it's due to the DNS issue, we could add a warning, but detect this on the _daemon_ side and return the warning over the API (similar to https://github.com/moby/moby/pull/38382)

It's better to validate on the daemon side, so that any client using the API would get the warning, and also because sometimes warnings are dependent on the version of the daemon (engine) in use

**Comment by KowKiller** [comment]
Hi,
I've experienced a similar problem in the past and recently I had the occasion to do [further analysis on it](https://github.com/moby/moby/issues/35488#issuecomment-482634553).

In my scenario, the containers connected to a docker network with a name longer than 63 characters are able to ping each other, but they experience a big delay before the ping starts.
Upon further analysis, this delay seems to be due to a failed reverse DNS lookup attempt done by the ping command in order to get the full hostname that, in case of docker containers, seems to be `container_name.container_network`.

This failure is likely to be due to the [DNS limit of 63 characters on labels](https://tools.ietf.org/html/rfc1034#section-3.1). So probably adding a warning is a good idea. The network will still work but some functionalities won't.

**Comment by neersighted** [comment]
Related but not the same: https://github.com/moby/moby/issues/36402

**Comment by thaJeztah** [comment]
related;

- https://github.com/moby/moby/issues/33433
- https://github.com/moby/moby/issues/31772
